packages/controllers/src/permissions/README.md

```diff
@@ -0,0 +1,187 @@
+Unrestricted methods can always be called by anyone.
+Restricted methods require the requisite permission in order to be called.
+
+At any given moment, the `PermissionController` state tree describes the complete state of the permissions of all subjects known to the current MetaMask instance.
+The `PermissionController` also provides methods for adding, updating, and removing permissions, and a `json-rpc-engine` middleware function factory so that the permissioning logic can be applied to incoming JSON-RPC requests.
+
+Once the permission middleware is injected into our middleware stack, every JSON-RPC request will be handled in one of the following ways:
+
+- If the requested method is neither restricted nor unrestricted, the request will be rejected with a `methodNotFound` error.
+- If the requested method is unrestricted, it will pass through the middleware unmodified.
+- If the requested method is restricted, the middleware will attempt to get the permission corresponding to the subject and target, and:
+  - If a permission is found, call the corresponding method with the request parameters.
+  - If a permission is not found, reject the request with an `unauthorized` error.
+
+### Caveats
+
+Caveats are objects associated with permissions that impose arbitrary restrictions on restricted method requests.
+Every permission has a `caveats` field, which is either an array of caveats or `null`.
+Every caveat has a string `type`, and every type has an associated function that is used to apply the caveat to a restricted method request.
+When the `PermissionController` is constructed, the consumer specifies the available caveat types and their implementations.
+
+> In `rpc-cap`, caveat types were specified by `rpc-cap` itself, and the consumer had to choose from the available types.
+> This did not end up being easy to reason about or work with in practice.
+> Now, caveat types and their implementations are provided by the consumer, just like the restricted methods and their implementations.
+
+## Examples
+
+### Construction
+
+```typescript
+// To construct a permission controller, we first need to define the caveat
+// types and restricted methods
+
+const caveatSpecifications = {
+  filterArrayResponse: {
+    type: 'filterArrayResponse',
+    // If a permission has any caveats, its corresponding restricted method
+    // implementation is decorated / wrapped with the implementations of its
+    // caveats, using the caveat's decorator function.
+    decorator: ((
+        // Restricted methods and other caveats can be async, so we have to
+        // assume that the method is async.
+        method: AsyncRestrictedMethod<Json, string[]>,
+        caveat: { type: 'filterArrayResponse'; value: string[] },
+      ) =>
+      async (args: RestrictedMethodOptions<RestrictedMethodParameters>) => {
+        const result = await method(args);
+        return Array.isArray(result)
```

Comment:
Nit: this should be updated to match the `filterArrayResponse` implementation in the tests, which seems generally better.